spec/history.cddl

```diff
@@ -0,0 +1,34 @@
+canister-change-entry = tagged<{
+  timestamp : timestamp
+  origin : canister-change-origin
+  change : canister-change
+}>
+
+canister-change-origin = {
+  from : "user"
```

Comment:
doesn't the `from: user | canister` field give a hint if a principal is which is the opposite of the design of not knowing what a principal is?

Reply:
I wouldn’t say so. The design behind a principal on its own is that it is opaque (as possible). An in some applications, we (or some of us) like to keep the difference hidden – e.g. update calls vs. inter-canister calls. But for certain use cases (like this one) the distinction is important, and then it’s ok to put that information next to the principal.

(It would be fishy if we wouldn’t have the user/canister tag here and instead say ”parse the principal”; _that_ would be in opposition to treating principals as opaque identifiers as far as possible.)

Reply:
Thanks for the clarification. It would be probably neat to add the keyword "(as possible)" in the docs (https://internetcomputer.org/docs/current/references/ic-interface-spec#principal). It makes the concept more clear in few words.